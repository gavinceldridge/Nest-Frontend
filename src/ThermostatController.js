import { React, useState, useEffect } from 'react'
import Form from "react-bootstrap/Form";
import Row from "react-bootstrap/Row";
import Col from "react-bootstrap/Col";
import Card from "react-bootstrap/Card";
import Button from "react-bootstrap/Button";
import Spinner from "react-bootstrap/Spinner";
import BackendApi from './BackendApi';

export default function ThermostatController() {

	const [formData, setFormData] = useState({ mode: "", modeOptions: ["OFF", "COOL", "HEAT"], temp: "", autodetect: "OFF" });
	// const [deviceInformation, setDeviceInformation] = useState();
	const [weatherInformation, setWeatherInformation] = useState([]);


	const convertCelsiusToFahrenheit = (cel) => {
		return (cel * (9 / 5) + 32).toFixed(1);
	}

	const updateWeatherInfo = async () => {
		setWeatherInformation([]);//to load the spinner
		const weatherInfo = await BackendApi.getWeatherInfo();
		console.log(weatherInfo);
		setWeatherInformation([weatherInfo.weather.main.feels_like, weatherInfo.weather.main.temp, weatherInfo.weather.main.humidity]);
	}

	const updateTemperature = (deviceInfo, mode) => {
		let temp = "";
		if (mode === "COOL") temp = convertCelsiusToFahrenheit(deviceInfo.devices[0].traits["sdm.devices.traits.ThermostatTemperatureSetpoint"]['coolCelsius']);
		else if (mode === "HEAT") temp = convertCelsiusToFahrenheit(deviceInfo.devices[0].traits["sdm.devices.traits.ThermostatTemperatureSetpoint"].heatCelsius);
		return temp;
	}

	const initData = async () => {

		const deviceInfo = await BackendApi.getDeviceInfo();
		// setDeviceInformation(deviceInfo.devices[0]);

		await updateWeatherInfo();
		const mode = deviceInfo.devices[0].traits["sdm.devices.traits.ThermostatMode"].mode;
		document.getElementById("mode").value = mode;
		let temp = updateTemperature(deviceInfo, mode);

		//get the current status of the smart detector
		const statusResponse = await BackendApi.changeTimer("");
		const status = statusResponse.status.toUpperCase();
		document.getElementById("autodetect").value = status;
		setFormData(data => ({
			...data,
			mode: mode,
			temp: temp,
			autodetect: status
		}));
	}

	//initialize UI data
	useEffect(() => {

		initData();

	}, []);

	const formChangeHandler = async (evt) => {

		// console.log(evt.target.value);
		const { name, value } = evt.target;

		console.log(name, value);
		setFormData(data => ({
			...data,
			[`${name}`]: value
		}));

		if (name === "mode") {
			await BackendApi.changeMode(value.toUpperCase());
		} else if (name === "autodetect") {
			await BackendApi.changeTimer("start");
		}
		if (name === "mode") {
			const deviceInfo = await BackendApi.getDeviceInfo();
			const mode = deviceInfo.devices[0].traits["sdm.devices.traits.ThermostatMode"].mode;
			const temp = await updateTemperature(deviceInfo, mode);
			setFormData(data => ({
				...data,
				temp
			}))
		}
	}

	const changeTemp = async () => {
		console.log(`changing temp to: ${formData.temp}`)
		await BackendApi.changeTemperature(formData.temp, formData.mode);
	}

	return (
		<div className="mt-3">
			<Row className="justify-content-center">
				<h1>Temperature Controller</h1>
			</Row>
			<Row className="justify-content-center mt-5">
				<Col xs={8}>

					<Form.Group as={Row} id="mode">
						<Form.Label column sm={5}><h3>Mode</h3></Form.Label>
						<Col sm={6}>
							<Form.Control as="select" size="lg" name="mode" onChange={formChangeHandler} value={formData.mode} >
								{formData.modeOptions.map(val => <option key={val}>{val}</option>)}
							</Form.Control>
						</Col>
					</Form.Group>

					<Form.Group as={Row} id="temp">
						<Form.Label column sm={5}><h3>Temperature</h3></Form.Label>
						<Col sm={5}>
							<Form.Control type="number" size="lg" name="temp" onChange={formChangeHandler} value={formData.temp}></Form.Control>
						</Col>
						<Col sm={1}>
							<Button variant="primary" name="confirm" onClick={changeTemp}>Confirm</Button>
						</Col>

					</Form.Group>

					<Form.Group as={Row} id="autodetect">
						<Form.Label column sm={5}><h3>Autodetect temp outside</h3></Form.Label>
						<Col sm={6}>
							<Form.Control as="select" size="lg" name="autodetect" onChange={formChangeHandler} value={formData.autodetect}>
								<option key={1}>ON</option>
								<option key={2}>OFF</option>
							</Form.Control>
						</Col>
					</Form.Group>
				</Col>
				<Col xs="8">
					<Row>
						<Col sm={5}>
							<h3>Weather Info</h3>
						</Col>
						<Col sm={6}>
							<Card>
								<Card.Header>
									<h5>Refresh: <Button variant="outline-primary" id="refreshBtn" onClick={updateWeatherInfo}><i className="fas fa-sync"></i></Button></h5>
								</Card.Header>
								<Card.Body>
									{weatherInformation.length === 3 ?
										<>
											<p>Feels Like: {weatherInformation[0]}</p>
											<p>Actual Temp: {weatherInformation[1]}</p>
											<p>Humidity: {weatherInformation[2]}</p>
										</>
										:
										<Spinner animation="border" />}

								</Card.Body>
							</Card>
						</Col>

					</Row>
				</Col>
			</Row >


		</div >
	)
}
