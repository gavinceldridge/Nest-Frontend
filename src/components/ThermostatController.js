import { React, useState, useEffect } from 'react'

import WeatherCard from './WeatherCard';
import UiPiece from './UiPiece';
import ModeSelector from './ModeSelector';

import Form from "react-bootstrap/Form";
import Row from "react-bootstrap/Row";
import Col from "react-bootstrap/Col";
import Button from "react-bootstrap/Button";
import BackendApi from '../BackendApi';

export default function ThermostatController() {

	const [formData, setFormData] = useState({ mode: "", modeOptions: ["OFF", "COOL", "HEAT"], temp: "", autodetect: "OFF" });
	// const [deviceInformation, setDeviceInformation] = useState();
	const [weatherInformation, setWeatherInformation] = useState([]);


	const convertCelsiusToFahrenheit = (cel) => {
		return (cel * (9 / 5) + 32).toFixed(1);
	}

	const updateWeatherInfo = async () => {
		setWeatherInformation([]);//to load the spinner
		const weatherInfo = await BackendApi.getWeatherInfo();
		console.log(weatherInfo);
		setWeatherInformation([weatherInfo.weather.main.feels_like, weatherInfo.weather.main.temp, weatherInfo.weather.main.humidity]);
	}

	const updateTemperature = (deviceInfo, mode) => {
		let temp = "";
		if (mode === "COOL") temp = convertCelsiusToFahrenheit(deviceInfo.devices[0].traits["sdm.devices.traits.ThermostatTemperatureSetpoint"]['coolCelsius']);
		else if (mode === "HEAT") temp = convertCelsiusToFahrenheit(deviceInfo.devices[0].traits["sdm.devices.traits.ThermostatTemperatureSetpoint"].heatCelsius);
		return temp;
	}

	const initData = async () => {

		const deviceInfo = await BackendApi.getDeviceInfo();
		// setDeviceInformation(deviceInfo.devices[0]);

		await updateWeatherInfo();
		const mode = deviceInfo.devices[0].traits["sdm.devices.traits.ThermostatMode"].mode;
		document.getElementById("mode").value = mode;
		let temp = updateTemperature(deviceInfo, mode);

		//get the current status of the smart detector
		const statusResponse = await BackendApi.changeTimer("");
		const status = statusResponse.status.toUpperCase();
		document.getElementById("autodetect").value = status;
		setFormData(data => ({
			...data,
			mode: mode,
			temp: temp,
			autodetect: status
		}));
	}

	//initialize UI data
	useEffect(() => {

		initData();

	}, []);

	const formChangeHandler = async (evt) => {

		// console.log(evt.target.value);
		const { name, value } = evt.target;

		console.log(name, value);
		setFormData(data => ({
			...data,
			[`${name}`]: value
		}));

		if (name === "mode") {
			await BackendApi.changeMode(value.toUpperCase());
		} else if (name === "autodetect") {
			await BackendApi.changeTimer("start");
		}
		if (name === "mode") {
			const deviceInfo = await BackendApi.getDeviceInfo();
			const mode = deviceInfo.devices[0].traits["sdm.devices.traits.ThermostatMode"].mode;
			const temp = await updateTemperature(deviceInfo, mode);
			setFormData(data => ({
				...data,
				temp
			}))
		}
	}

	const changeTemp = async () => {
		console.log(`changing temp to: ${formData.temp}`)
		await BackendApi.changeTemperature(formData.temp, formData.mode);
	}

	return (
		<div className="mt-3">
			<Row className="justify-content-center">
				<h1>Temperature Controller</h1>
			</Row>
			<Row className="justify-content-center mt-5">
				<Col xs={8}>

					<UiPiece formChangeHandler={formChangeHandler} formData={formData} />

					<Form.Group as={Row} id="temp">
						<Form.Label column sm={5}><h3>Temperature</h3></Form.Label>
						<Col sm={5}>
							<Form.Control type="number" size="lg" name="temp" onChange={formChangeHandler} value={formData.temp}></Form.Control>
						</Col>
						<Col sm={1}>
							<Button variant="primary" name="confirm" onClick={changeTemp}>Confirm</Button>
						</Col>

					</Form.Group>

					<Form.Group as={Row} id="autodetect">
						<Form.Label column sm={5}><h3>Autodetect temp outside</h3></Form.Label>
						<Col sm={6}>
							<Form.Control as="select" size="lg" name="autodetect" onChange={formChangeHandler} value={formData.autodetect}>
								<option key={1}>ON</option>
								<option key={2}>OFF</option>
							</Form.Control>
						</Col>
					</Form.Group>
				</Col>
				<Col xs="8">
					<Row>
						<Col sm={5}>
							<h3>Weather Info</h3>
						</Col>
						<Col sm={6}>
							<WeatherCard updateWeatherInfo={updateWeatherInfo} weatherInformation={weatherInformation} />
						</Col>

					</Row>
				</Col>
			</Row >


		</div >
	)
}
