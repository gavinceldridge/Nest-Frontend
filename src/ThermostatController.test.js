import { render, fireEvent } from '@testing-library/react';
import ThermostatController from './ThermostatController';
import React from "react";
import BackendApi from "./BackendApi";

//smoke test
test('renders without crashing', () => {
    render(<ThermostatController />);
});

//smoke test
test("render matches snapshot", () => {
    const { asFragment } = render(<ThermostatController />);
    expect(asFragment()).toMatchSnapshot();
});

//set the thermo mode to cool via an API call, then submit the temperature change form button, and verify temp through the API again. 
test("updates temperature upon submission", async () => {
    const { queryByTestId } = render(<ThermostatController />);
    // expect(queryByTestId("confirmBtn")).toBeInTheDocument();
    expect(document.getElementsByName("temp")[0]).toBeInTheDocument();
    expect(document.getElementsByName("confirm")[0]).toBeInTheDocument();

    //set the thermostat mode to cool:
    await BackendApi.changeMode("COOL");

    document.getElementsByName("temp")[0].value = 70;
    fireEvent.click(document.getElementsByName("confirm")[0]);

    //wait 5 seconds to ensure the temp change request was fulfilled
    setTimeout(async () => {

        const result = await BackendApi.getDeviceInfo();
        console.log(result);

    }, 5000);
});
