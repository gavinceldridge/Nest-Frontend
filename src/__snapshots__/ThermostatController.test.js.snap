// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="mt-3"
  >
    <div
      class="justify-content-center row"
    >
      <h1>
        Temperature Controller
      </h1>
    </div>
    <div
      class="justify-content-center mt-5 row"
    >
      <div
        class="col-8"
      >
        <div
          class="form-group row"
          id="mode"
        >
          <label
            class="form-label col-form-label col-sm-5"
          >
            <h3>
              Mode
            </h3>
          </label>
          <div
            class="col-sm-6"
          >
            <select
              class="form-control form-control-lg"
              name="mode"
            >
              <option>
                OFF
              </option>
              <option>
                COOL
              </option>
              <option>
                HEAT
              </option>
            </select>
          </div>
        </div>
        <div
          class="form-group row"
          id="temp"
        >
          <label
            class="form-label col-form-label col-sm-5"
          >
            <h3>
              Temperature
            </h3>
          </label>
          <div
            class="col-sm-5"
          >
            <input
              class="form-control form-control-lg"
              name="temp"
              type="number"
              value=""
            />
          </div>
          <div
            class="col-sm-1"
          >
            <button
              class="btn btn-primary"
              name="confirm"
              type="button"
            >
              Confirm
            </button>
          </div>
        </div>
        <div
          class="form-group row"
          id="autodetect"
        >
          <label
            class="form-label col-form-label col-sm-5"
          >
            <h3>
              Autodetect temp outside
            </h3>
          </label>
          <div
            class="col-sm-6"
          >
            <select
              class="form-control form-control-lg"
              name="autodetect"
            >
              <option>
                ON
              </option>
              <option>
                OFF
              </option>
            </select>
          </div>
        </div>
      </div>
      <div
        class="col-8"
      >
        <div
          class="row"
        >
          <div
            class="col-sm-5"
          >
            <h3>
              Weather Info
            </h3>
          </div>
          <div
            class="col-sm-6"
          >
            <div
              class="card"
            >
              <div
                class="card-header"
              >
                <h5>
                  Refresh: 
                  <button
                    class="btn btn-outline-primary"
                    id="refreshBtn"
                    type="button"
                  >
                    <i
                      class="fas fa-sync"
                    />
                  </button>
                </h5>
              </div>
              <div
                class="card-body"
              >
                <div
                  class="spinner-border"
                />
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
